### PR TITLE
Serialization for diff and profile tabs.

### PR DESCRIPTION
**Pull request by polina-c**

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 1]
```diff
@@ -26,19 +26,62 @@ import '../data/csv.dart';
...
+      core._snapshots.value.addAll(snapshots);
+    }
+    derived._updateValues();
+  }
 
   factory DiffPaneController.fromJson(Map<String, dynamic> json) {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return DiffPaneController(loader: null);
+    final snapshots = (json[_Json.snapshots] as List)
+        .map((e) => SnapshotDataItem.fromJson(e))
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:49]: I think this is using dynamic invocation in the `map` call. `e` is dynamic here. To avoid this, you can cast before calling .map

`(json[_Json.snaphots] as List).cast<Map<String, dynamic>>().map(...).toList()`

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 2]
```diff
@@ -26,19 +26,62 @@ import '../data/csv.dart';
...
+        .toList();
+
+    final diffWith = json[_Json.diffWith] as List<int?>;
+
+    for (var i = 0; i < snapshots.length; i++) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:54]: i can be final

Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:54]: before this loop can we assert that `snapshots.length` == `diffWith.length`? otherwise we may hit index out of range errors in the loop

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_control_pane.dart [hunk 3]
```diff
@@ -76,7 +76,7 @@ class _DiffDropdown extends StatelessWidget {
 
   List<DropdownMenuItem<SnapshotDataItem>> items() =>
       controller.core.snapshots.value
-          .where((item) => item.hasData)
+          .where((item) => item is SnapshotDataItem && item.isProcessed)
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_control_pane.dart:79]: you can use `whereType` here so that `item` will already be typed properly

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 4]
```diff
@@ -26,19 +26,62 @@ import '../data/csv.dart';
...
-    return {};
+    final snapshots = core.snapshots.value
+        .whereType<SnapshotDataItem>()
+        .where((s) => s.heap != null)
+        .toList();
+
+    final snapshotToIndex =
+        snapshots.asMap().map((index, item) => MapEntry(item, index));
+
+    final diffsWith = snapshots.map((item) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:76]: nit rename to `diffWithSnapshots`

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/snapshot_item.dart [hunk 5]
```diff
@@ -11,34 +13,56 @@ import '../../../../../shared/memory/heap_graph_loader.dart';
...
+      nameOverride: json[_Json.nameOverride] as String?,
+    );
+
+    final loader = HeapGraphLoaderFromChunks(
+      chunks: json[_Json.chunks] as List<ByteData>,
+      created: json[_Json.created] as DateTime,
+    );
+
+    unawaited(
+      result.loadHeap(loader),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/snapshot_item.dart:51]: formatting: remove trailing comma and then put comment above this line

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/snapshot_item.dart [hunk 6]
```diff
@@ -11,34 +13,56 @@ import '../../../../../shared/memory/heap_graph_loader.dart';
...
+    ); // Start the loading process, that will result in progress indicator in UI.
+
+    return result;
+  }
+
+  Map<String, dynamic> toJson() {
+    final heap = _heap!; // Not processed heaps are not serializable.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/snapshot_item.dart:58]: we do not want to throw exceptions at runtime. Should return an empty map instead?

File: packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart [hunk 7]
```diff
@@ -455,12 +455,6 @@ class AllocationProfileTableView extends StatefulWidget {
 
 class AllocationProfileTableViewState
     extends State<AllocationProfileTableView> {
-  @override
-  void didChangeDependencies() {
-    super.didChangeDependencies();
-    widget.controller.initialize();
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart:461]: this is changing the lifecycle of when we pay the cost to initialize the `ProfilePaneController`. The way this code used to be delayed controller initialization until we actually needed it when the Allocation profile view was opened. Are we certain this is what we want to do?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 8]
```diff
@@ -26,19 +26,62 @@ import '../data/csv.dart';
...
+      final diffWith = item.diffWith.value;
+      return diffWith == null ? null : snapshotToIndex[diffWith];
+    }).toList();
+
+    return {
+      _Json.snapshots: snapshots,
+      _Json.diffWith: diffsWith,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:83]: is this as space efficient as it can be? Or are we storing snapshots twice if there is a diff with value?

File: packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart [hunk 9]
```diff
@@ -44,3 +46,16 @@ class HeapGraphLoaderFile implements HeapGraphLoader {
     );
   }
 }
+
+class HeapGraphLoaderFromChunks implements HeapGraphLoader {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart:50]: dartdoc

File: packages/devtools_app/lib/src/shared/memory/heap_data.dart [hunk 10]
```diff
@@ -12,25 +14,28 @@ import 'retainers.dart';
...
-  HeapData._(
-    this.graph,
-    this.classes,
-    this.footprint, {
+  HeapData(
+    this.graph, {
     required this.created,
-    required this.retainedSizes,
-  });
+    @visibleForTesting bool startCalculation = true,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_data.dart:21]: this is unused, so you can remove

**Review by kenzieschmoll — COMMENTED** [review]
Please add tests for the new serialization and any other affected code as part of this change.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart (hunk 2 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:54]: Screenshot 2024-05-07 at 2 48 53 PM

File: packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart (hunk 7 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart:461]: It is job of memory controller, not widget, to initiate initialization of ProfilePaneController.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart (hunk 8 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:83]: diffsWith is just array of indexes

File: packages/devtools_app/lib/src/shared/memory/heap_data.dart (hunk 10 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/memory/heap_data.dart:21]: Added comment.

File: packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart (hunk 7 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/screens/memory/panes/profile/profile_view.dart:461]: Performing the initialization in the widget lifecycle makes this lazy, which we probably still want. Otherwise we will pay the cost of listeners firing in the background for a feature that may not even be in use. This is similar to how we don't initialize the memory controller if we are on the flutter inspector, for example. We initialize on first use.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:57]: Oh whoops. I was thinking this was `for (final foo in someIterable)`.

File: packages/devtools_app/lib/src/screens/memory/shared/primitives/memory_timeline.dart [hunk 11]
```diff
@@ -123,7 +123,7 @@ class MemoryTimeline {
     _eventSamples.add(
       EventSample.gcEvent(
         timestamp,
-        events: extensionEvents,
+        theEvents: extensionEvents,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/primitives/memory_timeline.dart:126]: rename to events

File: packages/devtools_shared/lib/src/memory/event_sample.dart [hunk 12]
```diff
@@ -163,12 +163,12 @@ class EventSample {
     this.extensionEvents,
   );
 
-  EventSample.gcEvent(this.timestamp, {ExtensionEvents? events})
+  EventSample.gcEvent(this.timestamp, {ExtensionEvents? theEvents})
```
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/memory/event_sample.dart:166]: rename to events. The in scope var should be used.

File: packages/devtools_app/lib/src/screens/memory/shared/primitives/memory_timeline.dart (hunk 11 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/shared/primitives/memory_timeline.dart:126]: Then it will conflict with the field name

File: packages/devtools_shared/lib/src/memory/event_sample.dart (hunk 12 above)
Comment (polina-c) [reply packages/devtools_shared/lib/src/memory/event_sample.dart:166]: Then it will conflict with the field name

**Comment by polina-c** [comment]
> Please add tests for the new serialization and any other affected code as part of this change.

Added tests for serialization.
Most of other code is test covered.

I will add more coverage after completing the feature and before  removing the flag.

File: packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart [hunk 13]
```diff
@@ -25,31 +26,29 @@ class OfflineMemoryData {
...
-        json[key] as Map<String, dynamic>? ?? {};
     return OfflineMemoryData(
-      DiffPaneController.fromJson(item(_Json.diffData)),
-      ProfilePaneController.fromJson(item(_Json.profileData)),
-      ChartData.fromJson(item(_Json.chartData)),
-      ClassFilter.fromJson(item(_Json.classFilter)),
+      deserialize(json[_Json.diffData], DiffPaneController.fromJson),
+      deserialize(json[_Json.profileData], ProfilePaneController.fromJson),
+      deserialize(json[_Json.chartData], ChartData.fromJson),
+      deserialize(json[_Json.classFilter], ClassFilter.fromJson),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart:33]: add the generic type argument to `deserialize`

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart [hunk 14]
```diff
@@ -38,11 +38,15 @@ class ChartData {
   }
 
   factory ChartData.fromJson(Map<String, dynamic> json) {
+    final timeline = json[_Json.timeline];
     final result = ChartData(
       mode: ControllerCreationMode.offlineData,
       isDeviceAndroid: json[_Json.isDeviceAndroid] as bool? ?? false,
-      timeline:
-          MemoryTimeline.fromJson(json[_Json.timeline] as Map<String, dynamic>),
+      timeline: timeline is MemoryTimeline
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:45]: can you use the deserialize helper here instead?

File: packages/devtools_app/lib/src/shared/primitives/serialization.dart [hunk 15]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+typedef FromJson<T> = T Function(Map<String, dynamic> json);
+
+/// Deserialized an object if it is serialized.
+// ignore: avoid-dynamic, serialization is exception for the rule.
+T deserialize<T>(dynamic json, FromJson<T> deserializer) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/serialization.dart:9]: add tests for this helper.

File: packages/devtools_app/lib/src/shared/primitives/serialization.dart [hunk 16]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/serialization.dart:1]: 2024

File: packages/devtools_app/test/memory/chart/controller/chart_data_test.dart [hunk 17]
```diff
@@ -0,0 +1,32 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/chart/controller/chart_data_test.dart:1]: 2024

File: packages/devtools_app/test/memory/chart/controller/chart_data_test.dart [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  test(
+    '$ChartData serializes and deserializes correctly',
+    () {
+      final item = ChartData(
+        mode: ControllerCreationMode.offlineData,
+        isDeviceAndroid: true,
+        timeline: MemoryTimeline(),
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/chart/controller/chart_data_test.dart:18]: can you add a test with non-empty timeline data?

File: packages/devtools_app/test/memory/chart/controller/chart_data_test.dart [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        interval: ChartInterval.theDefault,
+        isLegendVisible: true,
+      );
+
+      final json = item.toJson();
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/chart/controller/chart_data_test.dart:23]: assert the value of this:

```
expect(item.toJson(), {...});
```

File: packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart [hunk 20]
```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart:3]: remove duplicate copyright

File: packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    '$SnapshotDataItem serializes and deserializes correctly',
+    () async {
+      final item = SnapshotDataItem(
+        defaultName: 'defaultName',
+        displayNumber: 1,
+        nameOverride: 'nameOverride',
+      );
+      await item.loadHeap(HeapGraphLoaderGoldens());
+
+      final json = item.toJson();
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart:21]: assert the values created by toJson to verify expectations

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 22]
```diff
@@ -19,26 +19,72 @@ import '../../../../../shared/memory/classes.dart';
...
+      core._snapshots.addAll(snapshots);
+    }
+    derived._updateValues();
+  }
 
   factory DiffPaneController.fromJson(Map<String, dynamic> json) {
-    // TODO(polina-c): implement, https://github.com/flutter/devtools/issues/6972
-    return DiffPaneController(loader: null);
+    final snapshots = (json[_Json.snapshots] as List)
+        .map((e) => deserialize(e, SnapshotDataItem.fromJson))
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:50]: add type to deserialize

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 23]
```diff
@@ -19,26 +19,72 @@ import '../../../../../shared/memory/classes.dart';
...
-    return {};
+    final snapshots = core.snapshots.value
+        .whereType<SnapshotDataItem>()
+        .where((s) => s.heap != null)
+        .toList();
+
+    final snapshotToIndex =
+        snapshots.asMap().map((index, item) => MapEntry(item, index));
+
+    final diffsWithSnapshots = snapshots.map((item) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:79]: nit: now that I better understand what this var is, how about "diffWithIndices" and then below r/diffWith/diffWithIndex

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 24]
```diff
@@ -19,26 +19,72 @@ import '../../../../../shared/memory/classes.dart';
...
+      final diffWith = item.diffWith.value;
+      return diffWith == null ? null : snapshotToIndex[diffWith];
+    }).toList();
+
+    return {
+      _Json.snapshots: snapshots,
+      _Json.diffWith: diffsWithSnapshots,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:86]: same here: _Json.diffWithIndices

File: packages/devtools_app/lib/src/shared/memory/heap_data.dart [hunk 25]
```diff
@@ -12,25 +14,27 @@ import 'retainers.dart';
...
 /// Raw and calculated data of the heap snapshot.
-@immutable
 class HeapData {
-  HeapData._(
-    this.graph,
-    this.classes,
-    this.footprint, {
+  HeapData(
+    this.graph, {
     required this.created,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_data.dart:20]: nit: this will all fit on one line so no need for trailing comma

File: packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart [hunk 26]
```diff
@@ -44,3 +46,17 @@ class HeapGraphLoaderFile implements HeapGraphLoader {
...
   }
 }
+
+/// Loads a heap snapshot from `List<ByteData>` created with HeapSnapshotGraph.toChunks.
+class HeapGraphLoaderFromChunks implements HeapGraphLoader {
+  HeapGraphLoaderFromChunks({required this.chunks, required this.created});
+
+  List<ByteData> chunks;
+
+  DateTime created;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart:56]: should these be final?

File: packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart [hunk 27]
```diff
@@ -44,3 +46,17 @@ class HeapGraphLoaderFile implements HeapGraphLoader {
...
+
+  @override
+  Future<(HeapSnapshotGraph, DateTime)> load() async {
+    return (HeapSnapshotGraph.fromChunks(chunks), created);
+  }
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart:61]: this doesn't need to be async. If the super class returns FutureOr, then we can remove the unnecessary async here

File: packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+      final controller = scene.controller.diff;
+
+      final snapshots =
+          controller.core.snapshots.value.whereType<SnapshotDataItem>();
+
+      expect(snapshots.length, 2);
+      snapshots.first.diffWith.value = snapshots.last;
+
+      final fromJson = DiffPaneController.fromJson(controller.toJson());
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart:56]: can you assert the content of .toJson() and .fromJson() so we can verify the values of the serialization are as expected?

File: packages/devtools_app/test/memory/framework/offline_data_test.dart [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+          isDeviceAndroid: true,
+          timeline: MemoryTimeline(),
+          interval: ChartInterval.theDefault,
+          isLegendVisible: true,
+        ),
+        ClassFilter.empty(),
+        selectedTab: 0,
+      );
+
+      final json = item.toJson();
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/framework/offline_data_test.dart:33]: assert values of toJson result

File: packages/devtools_shared/lib/src/memory/event_sample.dart (hunk 12 above)
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/memory/event_sample.dart:166]: you can use `this.events = events`

File: packages/devtools_shared/lib/src/memory/heap_sample.dart [hunk 30]
```diff
@@ -53,9 +53,9 @@ class HeapSample {
         'used': used,
         'external': external,
         'gc': isGC,
-        'adb_memoryInfo': adbMemoryInfo.toJson(),
-        'memory_eventInfo': memoryEventInfo.toJson(),
-        'raster_cache': rasterCache.toJson(),
+        'adb_memoryInfo': adbMemoryInfo,
+        'memory_eventInfo': memoryEventInfo,
+        'raster_cache': rasterCache,
```
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/memory/heap_sample.dart:58]: I recommend that we don't make unnecessary changes to devtools_shared that aren't test covered. I don't think this code has good test coverage to verify nothing is broken here.

File: packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart (hunk 21 above)
Comment (polina-c) [reply packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart:21]: IMHO, it will increase maintenance cost of the test, without adding value.
We do not care what is in json, we just want it to produce the same values when deserialized, right?

File: packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart (hunk 28 above)
Comment (polina-c) [reply packages/devtools_app/test/memory/diff/controller/diff_pane_controller_test.dart:56]: about `toJson` see https://github.com/flutter/devtools/pull/7699#discussion_r1596044223

added some validation of `from`

File: packages/devtools_app/test/memory/framework/offline_data_test.dart (hunk 29 above)
Comment (polina-c) [reply packages/devtools_app/test/memory/framework/offline_data_test.dart:33]: see https://github.com/flutter/devtools/pull/7699#discussion_r1596044223

File: packages/devtools_shared/lib/src/memory/heap_sample.dart (hunk 30 above)
Comment (polina-c) [reply packages/devtools_shared/lib/src/memory/heap_sample.dart:58]: Added test coverage. Heap sample is part of chart data, stat needs to be serialized.

File: packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart (hunk 21 above)
Comment (kenzieschmoll) [reply packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart:21]: If the `.toJson()` value is not what we expect (like a field is missing for example or a default value is wrong), bugs could still slip in. For example both `.toJson()` and `fromJson()` could be missing a field.

File: packages/devtools_shared/lib/devtools_shared.dart [hunk 31]
```diff
@@ -16,4 +16,5 @@ export 'src/utils/compare.dart';
 export 'src/utils/file_utils.dart';
 export 'src/utils/retry.dart';
 export 'src/utils/semantic_version.dart';
+export 'src/utils/serialization.dart';
```
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/devtools_shared.dart:19]: add a changelog entry for changes to public package devtools_shared

File: packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart (hunk 21 above)
Comment (polina-c) [reply packages/devtools_app/test/memory/diff/controller/snapshot_item_test.dart:22]: I added validation that all keys are used. I skipped validation for values. 
Yes, bugs still can squeeze, but no test coverage is full. I believe verification for values is overhead in this case.

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart [hunk 32]
```diff
@@ -9,11 +9,15 @@ import '../../../../../shared/primitives/simple_items.dart';
...
-class _Json {
+// ignore: avoid_classes_with_only_static_members, enum-like classes are ok
+@visibleForTesting
+class Json {
   static const isDeviceAndroid = 'isAndroid';
   static const timeline = 'timeline';
   static const interval = 'interval';
   static const isLegendVisible = 'isLegendVisible';
+
+  static final all = [isDeviceAndroid, timeline, interval, isLegendVisible];
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:20]: make this a set for consistency above. Or preferably, just switch to using an enum so that you can use `Json.values`.

If you leave the `all` field, mark visible for testing if that is the only place it is used. Here and elsewhere.

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart [hunk 33]
```diff
@@ -11,13 +11,15 @@ import '../data/primitives.dart';
...
-class Json {
-  static const isDeviceAndroid = 'isAndroid';
-  static const timeline = 'timeline';
-  static const interval = 'interval';
-  static const isLegendVisible = 'isLegendVisible';
+enum Json {
+  isDeviceAndroid('isAndroid'),
+  timeline('timeline'),
+  interval('interval'),
+  isLegendVisible('isLegendVisible');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:18]: instead of passing the string representation to each of these, consider this pattern where the constructor has an optional name override, and then you don't have to pass in the string for the values where `Json.enumValue.name` is the equivalent. 

example: https://github.com/flutter/devtools/blob/master/packages/devtools_app/integration_test/test_infra/run/_test_app_driver.dart/#L417

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 34]
```diff
@@ -31,11 +31,13 @@ import 'snapshot_item.dart';
...
-  static const snapshots = 'snapshots';
-  static const diffWith = 'diffWith';
+enum Json {
+  snapshots('snapshots'),
+  diffWith('diffWith');
 
-  static final all = [snapshots, diffWith];
+  const Json(this.key);
+
+  final String key;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:40]: unnecessary. Json.snapshots.name and Json.diffWith.name are equivalent. This comment applies here and elsewhere to other enums.

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart [hunk 35]
```diff
@@ -9,17 +9,18 @@ import '../../../../../shared/primitives/simple_items.dart';
...
+  interval,
+  isLegendVisible;
 
-  const Json(this.key);
+  const Json({String? nameOverride}) : _nameOverride = nameOverride;
 
-  final String key;
+  final String? _nameOverride;
+
+  String get key => _nameOverride ?? name;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:23]: since the override is not used, we don't need any of this.

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart [hunk 36]
```diff
@@ -16,11 +16,7 @@ enum Json {
   interval,
   isLegendVisible;
 
-  const Json({String? nameOverride}) : _nameOverride = nameOverride;
-
-  final String? _nameOverride;
-
-  String get key => _nameOverride ?? name;
+  String get key => name;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:19]: why not just use `.name` directly? Seems like added complexity to add a getter that just returns the unchanged name value.

File: packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart [hunk 37]
```diff
@@ -2,17 +2,29 @@
...
+import 'package:meta/meta.dart';
+
 import '../../panes/chart/controller/chart_data.dart';
 import '../../panes/diff/controller/diff_pane_controller.dart';
 import '../../panes/profile/profile_pane_controller.dart';
 import '../../shared/heap/class_filter.dart';
 
-class _Json {
+@visibleForTesting
+class Json {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/framework/offline_data/offline_data.dart:14]: make this an enum too

File: packages/devtools_shared/lib/src/memory/heap_sample.dart [hunk 38]
```diff
@@ -4,9 +4,37 @@
...
 import 'dart:convert';
 
+import 'package:meta/meta.dart';
+
+import '../utils/serialization.dart';
 import 'adb_memory_info.dart';
 import 'event_sample.dart';
 
+@visibleForTesting
+class Json {
```
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/memory/heap_sample.dart:14]: make this an enum

File: packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart (hunk 36 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart:19]: For me it is alert for future maintainer that key may be different from name, and it is needed for backward compatibility.
But, yes, it is debatable. 
Removed.

**Review by kenzieschmoll — APPROVED** [review]

File: packages/devtools_shared/CHANGELOG.md [hunk 39]
```diff
@@ -1,10 +1,14 @@
 # 10.0.0-dev.2
 * Support detecting package roots for nested Dart projects in the
 `packageRootFromFileUriString` utility method.
+* Added helper `deserialize`
+* Extended serialization for `HeapSample` and `ExtensionEvents`
```
Comment (kenzieschmoll) [inline packages/devtools_shared/CHANGELOG.md:5]: 10.0.0-dev.2 is already published. Please add a new section for dev.3 and bump the pubspec version. Then add these entires to the dev.3 section

File: packages/devtools_shared/CHANGELOG.md [hunk 40]
```diff
@@ -1,10 +1,14 @@
...
 
 # 10.0.0-dev.1
 * Export `DTDConnectionInfo` from the `devtools_shared.dart` library instead
 of the `devtools_server.dart` library.
+* Support detecting package roots for nested Dart projects in the
+`packageRootFromFileUriString` utility method.
```
Comment (kenzieschmoll) [inline packages/devtools_shared/CHANGELOG.md:11]: remove this added entry. This is part of dev.2 above
